Errorno:EBADF

I got errors like this.

```
[ERROR] Failed to execute goal jp.pizzafactory.maven:atk2abrex-maven-plugin:0.0.1-SNAPSHOT:make-csv (csv generation) on project atk2abrex-maven-plugin-test: Execution csv generation of goal jp.pizzafactory.maven:atk2abrex-maven-plugin:0.0.1-SNAPSHOT:make-csv failed: org.jruby.exceptions.RaiseException: (Errno::EBADF) Bad file descriptor -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal jp.pizzafactory.maven:atk2abrex-maven-plugin:0.0.1-SNAPSHOT:make-csv (csv generation) on project atk2abrex-maven-plugin-test: Execution csv generation of goal jp.pizzafactory.maven:atk2abrex-maven-plugin:0.0.1-SNAPSHOT:make-csv failed: org.jruby.exceptions.RaiseException: (Errno::EBADF) Bad file descriptor
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:224)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution csv generation of goal jp.pizzafactory.maven:atk2abrex-maven-plugin:0.0.1-SNAPSHOT:make-csv failed: org.jruby.exceptions.RaiseException: (Errno::EBADF) Bad file descriptor
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	... 20 more
Caused by: org.jruby.embed.InvokeFailedException: org.jruby.exceptions.RaiseException: (Errno::EBADF) Bad file descriptor
	at org.jruby.embed.internal.EmbedRubyObjectAdapterImpl.callMethod(EmbedRubyObjectAdapterImpl.java:255)
	at org.jruby.embed.ScriptingContainer.callMethod(ScriptingContainer.java:1458)
	at jp.pizzafactory.maven.atk2abrex.MakeCsvMojo.execute(MakeCsvMojo.java:56)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	... 21 more
Caused by: org.jruby.exceptions.RaiseException: (Errno::EBADF) Bad file descriptor
	at org.jruby.RubyIO.pos(org/jruby/RubyIO.java:1651)
	at RUBY.current_line(/home/monaka/.m2/repository/org/jruby/jruby-complete/1.7.23/jruby-complete-1.7.23.jar!/META-INF/jruby.home/lib/ruby/1.9/rexml/source.rb:236)
	at RUBY.line(/home/monaka/.m2/repository/org/jruby/jruby-complete/1.7.23/jruby-complete-1.7.23.jar!/META-INF/jruby.home/lib/ruby/1.9/rexml/parseexception.rb:43)
	at RUBY.to_s(/home/monaka/.m2/repository/org/jruby/jruby-complete/1.7.23/jruby-complete-1.7.23.jar!/META-INF/jruby.home/lib/ruby/1.9/rexml/parseexception.rb:28)
	at org.jruby.RubyException.message(org/jruby/RubyException.java:157)
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

--- a/atk2abrex-maven-plugin-test/pom.xml
+++ b/atk2abrex-maven-plugin-test/pom.xml
@@ -9,12 +9,34 @@
 	<build>
 		<plugins>
 			<plugin>
+				<groupId>com.googlecode.maven-download-plugin</groupId>
+				<artifactId>download-maven-plugin</artifactId>
+			</plugin>
+			<plugin>
 				<groupId>jp.pizzafactory.maven</groupId>
 				<artifactId>atk2abrex-maven-plugin</artifactId>
 			</plugin>
 		</plugins>
 		<pluginManagement>
 			<plugins>
+				<plugin>
+					<groupId>com.googlecode.maven-download-plugin</groupId>
+					<artifactId>download-maven-plugin</artifactId>
+					<version>1.2.1</version>
+					<executions>
+						<execution>
+							<id>get-ecuc-param</id>
+							<phase>generate-sources</phase>
+							<goals>
+								<goal>wget</goal>
+							</goals>
+							<configuration>
+								<url>http://www.autosar.org/fileadmin/files/releases/4-0/methodology-templates/templates/standard/AUTOSAR_MOD_ECUConfigurationParameters.zip</url>
+								<unpack>true</unpack>
+							</configuration>
+						</execution>
+					</executions>
+				</plugin>
 				<plugin>
 					<groupId>jp.pizzafactory.maven</groupId>
 					<artifactId>atk2abrex-maven-plugin</artifactId>
@@ -42,6 +64,17 @@
 								<files>
 									<file>${basedir}/target/sample.arxml</file>
 								</files>
+							</configuration>
+						</execution>
+						<execution>
+							<id>csv generation</id>
+							<phase>generate-resources</phase>
+							<goals>
+								<goal>make-csv</goal>
+							</goals>
+							<configuration>
+								<ecuConfigurationParameters>${basedir}/target/AUTOSAR_MOD_ECUConfigurationParameters.arxml</ecuConfigurationParameters>
+								<targetModule>Com</targetModule>
 							</configuration>
 						</execution>
 					</executions>

--- a/atk2abrex-maven-plugin/pom.xml
+++ b/atk2abrex-maven-plugin/pom.xml
@@ -41,7 +41,7 @@
 		<dependency>
 			<groupId>org.jruby</groupId>
 			<artifactId>jruby-complete</artifactId>
-			<version>1.7.23</version>
+			<version>9.0.4.0</version>
 		</dependency>
 		<dependency>
 			<groupId>commons-io</groupId>

--- a/atk2abrex-maven-plugin/src/main/java/jp/pizzafactory/maven/atk2abrex/MakeCsvMojo.java
+++ b/atk2abrex-maven-plugin/src/main/java/jp/pizzafactory/maven/atk2abrex/MakeCsvMojo.java
@@ -1,0 +1,63 @@
+package jp.pizzafactory.maven.atk2abrex;
+
+/*
+ * Copyright 2016 PizzaFactory Project.
+ *
+ * Licensed under the EPL-1.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.eclipse.org/legal/epl-v10.html
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.File;
+import java.io.InputStream;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.jruby.Ruby;
+import org.jruby.embed.ScriptingContainer;
+import org.jruby.runtime.ThreadContext;
+
+/**
+ */
+@Mojo(name = "make-csv", defaultPhase = LifecyclePhase.PROCESS_SOURCES)
+public class MakeCsvMojo extends AbstractAtk2AbrexMojo {
+
+    /**
+     * Location of the source ARXML
+     */
+    @Parameter(required = true)
+    private File ecuConfigurationParameters;
+
+    /**
+     * Name of the target module.
+     */
+    @Parameter(required = true)
+    private String targetModule;
+
+    @Override
+    public void execute() throws MojoExecutionException {
+        Ruby ruby = Ruby.newInstance();
+        ThreadContext context = ThreadContext.newContext(ruby);
+        ScriptingContainer container = new ScriptingContainer();
+        InputStream ist = YamlToArxmlMojo.class.getResourceAsStream("abrex.rb");
+        container.runScriptlet(ist, "abrex.rb");
+
+        final String sOutputDir = outputDirectory.getAbsolutePath();
+        final String sFileName = ecuConfigurationParameters.getAbsolutePath();
+
+        container.callMethod(context, "MakeCsv", sOutputDir, sFileName,
+                targetModule);
+        container.terminate();
+    }
+
+}

--- a/atk2abrex-maven-plugin/src/main/resources/jp/pizzafactory/maven/atk2abrex/abrex.rb
+++ b/atk2abrex-maven-plugin/src/main/resources/jp/pizzafactory/maven/atk2abrex/abrex.rb
@@ -804,13 +804,13 @@ end
 ######################################################################
 # ジェネレータ用csvファイルの生成
 ######################################################################
-def MakeCsv(sFileName, sTargetModule)
+def MakeCsv(sOutputDir, sFileName, sTargetModule)
   # ファイルが存在しない場合エラー
   if (!File.exist?(sFileName))
     abort("Argument error !! [#{sFileName}]")
   end
 
-  sCsvFileName = File.dirname(sFileName) + "/" + sTargetModule + ".csv"
+  sCsvFileName = sOutputDir + "/" + sTargetModule + ".csv"
 
   # XMLライブラリでの読み込み
   cXmlData = REXML::Document.new(open(sFileName))

